### PR TITLE
Remove LegacyCTRAListener from MetadataUpdateSettingsService

### PR DESCRIPTION
**Pull request by arteam**
See #83784

### CODE REVIEW

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
I would prefer we clean up this class's task/executor as part of this change, bring it more in line with how we think these things should be implemented. That's really the purpose of the `LegacyClusterTaskResultActionListener` marker.

 E.g. the task itself should mostly just be data (typically a `record`, and definitely not a subclass of `AckedClusterStateUpdateTask`) and the execution logic moved to the executor.

Ideally we'd not build a new cluster state each time round the loop too, although I tried that with this task and it seemed pretty tricky because of how we update the `Metadata.Builder`  before we've finished doing all the validation steps. I can live without that change here, but we should definitely do more cleanup than just this one thing.

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by arteam** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsService.java [hunk 1]
```diff
@@ -87,183 +98,218 @@ public MetadataUpdateSettingsService(
...
+        }
+
+        @Override
+        public void onAllNodesAcked() {
+            listener.onResponse(AcknowledgedResponse.of(true));
+        }
+
+        @Override
+        public void onAckFailure(Exception e) {
+            listener.onResponse(AcknowledgedResponse.of(false));
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsService.java:286]: Just realised we have constants for these things:

```suggestion
            listener.onResponse(AcknowledgedResponse.FALSE);
```

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsService.java [hunk 2]
```diff
@@ -87,183 +98,218 @@ public MetadataUpdateSettingsService(
...
+        }
+
+        @Override
+        public boolean mustAck(DiscoveryNode discoveryNode) {
+            return true;
+        }
+
+        @Override
+        public void onAllNodesAcked() {
+            listener.onResponse(AcknowledgedResponse.of(true));
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsService.java:281]:
```suggestion
            listener.onResponse(AcknowledgedResponse.TRUE);
```

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsService.java [hunk 3]
```diff
@@ -87,183 +98,218 @@ public MetadataUpdateSettingsService(
...
+        }
+
+        @Override
+        public void clusterStateProcessed(ClusterState oldState, ClusterState newState) {
+            assert false : "not called";
+        }
+    }
+
+    public void updateSettings(final UpdateSettingsClusterStateUpdateRequest request, final ActionListener<AcknowledgedResponse> listener) {
+        ActionListener<AcknowledgedResponse> actionListener = wrapPreservingContext(listener, threadPool.getThreadContext());
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsService.java:306]: Hmm I think this wrapping is very old and hasn't been necessary for ages. Could you add an `ESSingleNodeTestCase` test which sets up a thread context, calls the update settings API, and asserts that the context is passed through to the listener (even without this wrapping). See e.g. how `MasterServiceTests#testTaskFailureNotification` interacts with the thread context.

**Review by DaveCTurner — COMMENTED** [review]
Looks good.

I now realise that we didn't add tests when we made this API implement batching. I expect any problems would trip something somewhere but it would be good to have a test suite that focusses on this. See `MetadataIndexStateServiceBatchingTests` for some ideas about how to write this kind of test.

**Pull request closed without merging**